nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java

```diff
@@ -0,0 +1,367 @@
+            if (osName.startsWith("Linux")) {
+                logger.debug("Selected Linux command set.");
+                commands = new NssShellCommands();
+            } else if (osName.startsWith("Mac OS X")) {
+                logger.debug("Selected OSX command set.");
+                commands = new OsxShellCommands();
+            } else {
+                throw new AuthorizerCreationException(OS_TYPE_ERROR);
+            }
+            setCommandsProvider(commands);
+        }
+
+        // Our second init step is to run the system check from that
+        // command set to determine if the other commands will work on
+        // this host or not.
+        try {
+            ShellRunner.runShell(commands.getSystemCheck());
+        } catch (final IOException ioexc) {
+            logger.error("initialize exception: " + ioexc + " system check command: " + commands.getSystemCheck());
+            throw new AuthorizerCreationException(SYS_CHECK_ERROR, ioexc.getCause());
+        }
+
+        // With our command set selected, and our system check passed,
+        // we can pull in the users and groups:
+        refreshUsers();
+        refreshGroups();
+
+        // Our last init step is to fire off the refresh threads:
+        Integer initialDelay = 30, fixedDelay = 30;
+        scheduler.scheduleWithFixedDelay(this::refreshUsers, initialDelay, fixedDelay, TimeUnit.SECONDS);
+        scheduler.scheduleWithFixedDelay(this::refreshGroups, initialDelay, fixedDelay, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Called immediately before instance destruction for implementers to release resources.
+     *
+     * @throws AuthorizerDestructionException If pre-destruction fails.
+     */
+    @Override
+    public void preDestruction() throws AuthorizerDestructionException {
+        try {
+            scheduler.shutdownNow();
+        } catch (final Exception ignored) {
+        }
+    }
+
+    public ShellCommandsProvider getCommandsProvider() {
+        return selectedShellCommands;
+    }
+
+    public void setCommandsProvider(ShellCommandsProvider commandsProvider) {
+        selectedShellCommands = commandsProvider;
+    }
+
+    private void refreshUsers() {
+        Map<String, User> byId = new HashMap<>();
+        Map<String, User> byName = new HashMap<>();
+        List<String> lines;
+
+        try {
+            lines = ShellRunner.runShell(selectedShellCommands.getUsersList());
+        } catch (final IOException ioexc)  {
+            logger.error("refreshUsers shell exception: " + ioexc);
+            return;
+        }
+
+        lines.forEach(line -> {
+                String[] record = line.split(":");
+                if (record.length > 1) {
+                    String name = record[0],
+                        id = record[1];
+                    if (name != null && id != null && !name.equals("") && !id.equals("")) {
+                        User user = new User.Builder().identity(name).identifier(id).build();
```

Comment:
I took the approach of re-creating the map after each run to avoid the additional logic required for tracking add/modify/delete changes to the three sets in question (user, group, group membership).